### PR TITLE
[gui] only close modal dialogs if we leave a media window

### PR DESCRIPTION
**Pull request by xhaggi**
As introduced by #6546 we close every active dialog if we leave/switch a media window. This also close some non-modal dialogs e.g. extended progress dialog, volumebar dialog etc. which are window independent. This PR will fix it by only closing active modal dialogs instead of all.

This PR also combines the 2 different methods to register a dialog as active dialog into one with a better wording.

@mkortstiege as discussed at IRC, you know it best ;)
@AlwinEsch ping

### CODE REVIEW

**Comment by mkortstiege** [comment]
Looks good .. obviously ;) jenkins build this please.

File: xbmc/guilib/GUIWindowManager.cpp [hunk 1]
```diff
@@ -810,6 +813,16 @@ void CGUIWindowManager::CloseDialogs(bool forceClose) const
   }
 }
 
+void CGUIWindowManager::CloseModalDialogs(bool forceClose) const
+{
+  CSingleLock lock(g_graphicsContext);
```
Comment (FernetMenta) [inline xbmc/guilib/GUIWindowManager.cpp]: this looks deadlock happy. i won't hold the lock for expensive operations like closing dialogs. why is this needed?

**Comment by Montellese** [comment]
The changes look good AFAICT.

Out of interest: What is the use case where a media window is closed while a modal dialog is being shown?

File: xbmc/guilib/GUIWindowManager.cpp (hunk 1 above)
Comment (Montellese) [inline xbmc/guilib/GUIWindowManager.cpp]: `CGUIWindowManager::CloseDialogs` does the same. This looks like a copy with an additional condition. We have to protect against modifications of the `m_activeDialogs` vector.

Comment (xhaggi) [inline xbmc/guilib/GUIWindowManager.cpp]: we use the same for CloseDialogs() just c/p from there.

**Comment by xhaggi** [comment]
@Montellese for example if you stay at the movie window with the video info dialog open and you want to switch to music or Live TV etc. by pushing a remote button or key (builtin).

File: xbmc/guilib/GUIWindowManager.cpp (hunk 1 above)
Comment (xhaggi) [inline xbmc/guilib/GUIWindowManager.cpp]: should we drop this on both?

Comment (FernetMenta) [inline xbmc/guilib/GUIWindowManager.cpp]: those methods should only be called by render thread, right? if this is true protection of m_activeDialogs is not needed.
an alternative would be to introduce a separate lock for activeDialogs, gfx lock is rather heavy for that purpose

Comment (xhaggi) [inline xbmc/guilib/GUIWindowManager.cpp]: After a second look every dialog related method (e.g. RemoveDialog, HasModalDialog etc.) use a graphical lock.
Changing this is out of scope of this PR IMO.

Comment (FernetMenta) [inline xbmc/guilib/GUIWindowManager.cpp]: > Changing this is out of scope of this PR IMO.

agree

**Comment by xhaggi** [comment]
any other objections?

**Comment by uNiversaI** [comment]
I think this caused http://trac.kodi.tv/ticket/15943

**Comment by xhaggi** [comment]
@uNiversaI why do do suspect this has something to do with the issue?

**Comment by uNiversaI** [comment]
@xhaggi Sorry. :blush:  its not this (confirmed its not this because recompiled back before this was added) I suspect its https://github.com/xbmc/xbmc/pull/6549 instead but need to recompile just before that was merged to confirm it is that.
